arwen/elrondapi/elrondei.go

```diff
@@ -627,3 +656,280 @@ func int64finish(context unsafe.Pointer, value int64) {
 	gasToUse := hostContext.GasSchedule().ElrondAPICost.Int64Finish
 	hostContext.UseGas(gasToUse)
 }
+
+//export executeOnSameContext
+func executeOnSameContext(
+	context unsafe.Pointer,
+	gasLimit int64,
+	addressOffset int32,
+	valueOffset int32,
+	functionOffset int32,
+	functionLength int32,
+	numArguments int32,
+	argumentsLengthOffset int32,
+	dataOffset int32,
+) int32 {
+	instCtx := wasmer.IntoInstanceContext(context)
+	ethContext := arwen.GetEthContext(instCtx.Data())
+
+	send := ethContext.GetSCAddress()
+	dest := arwen.LoadBytes(instCtx.Memory(), addressOffset, arwen.AddressLen)
+	value := arwen.LoadBytes(instCtx.Memory(), valueOffset, arwen.BalanceLen)
+	function, data, actualLen := getArgumentsFromMemory(context, functionOffset, functionLength, numArguments, argumentsLengthOffset, dataOffset)
+
+	gasToUse := ethContext.GasSchedule().ElrondAPICost.ExecuteOnSameContext
+	gasToUse += ethContext.GasSchedule().BaseOperationCost.StorePerByte * uint64(actualLen)
+	ethContext.UseGas(gasToUse)
+
+	if ethContext.GasLeft() < uint64(gasLimit) {
+		return 1
+	}
+
+	bigIntVal := big.NewInt(0).SetBytes(value)
+	ethContext.Transfer(dest, send, bigIntVal, nil)
+
+	contractCallInput := &vmcommon.ContractCallInput{
+		VMInput: vmcommon.VMInput{
+			CallerAddr:  send,
+			Arguments:   data,
+			CallValue:   bigIntVal,
+			GasPrice:    0,
+			GasProvided: uint64(gasLimit),
+		},
+		RecipientAddr: dest,
+		Function:      function,
+	}
+	err := ethContext.ExecuteOnDestContext(contractCallInput)
+	if err != nil {
+		return 1
+	}
+
+	return 0
+}
+
+//export executeOnDestContext
+func executeOnDestContext(
+	context unsafe.Pointer,
+	gasLimit int64,
+	addressOffset int32,
+	valueOffset int32,
+	functionOffset int32,
+	functionLength int32,
+	numArguments int32,
+	argumentsLengthOffset int32,
+	dataOffset int32,
+) int32 {
+	instCtx := wasmer.IntoInstanceContext(context)
+	ethContext := arwen.GetEthContext(instCtx.Data())
+
+	send := ethContext.GetSCAddress()
+	dest := arwen.LoadBytes(instCtx.Memory(), addressOffset, arwen.AddressLen)
+	value := arwen.LoadBytes(instCtx.Memory(), valueOffset, arwen.BalanceLen)
+	function, data, actualLen := getArgumentsFromMemory(context, functionOffset, functionLength, numArguments, argumentsLengthOffset, dataOffset)
+
+	gasToUse := ethContext.GasSchedule().ElrondAPICost.ExecuteOnDestContext
+	gasToUse += ethContext.GasSchedule().BaseOperationCost.StorePerByte * uint64(actualLen)
+	ethContext.UseGas(gasToUse)
+
+	if ethContext.GasLeft() < uint64(gasLimit) {
+		return 1
+	}
+
```

Comment:
Probably need `gasLimit := erdContext.GasLeft()` here.

Reply:
no, gasLimit is input for function